vault/logical_system.go

```diff
@@ -214,6 +226,360 @@ type SystemBackend struct {
+		}
+
+		// TODO: Should we delete after every attempt or allow multiple tries?
+		delete(pendingRaftPeers, serverID)
+
+		if subtle.ConstantTimeCompare(answer, expectedAnswer) == 0 {
+			return logical.ErrorResponse("invalid answer given"), logical.ErrInvalidRequest
+		}
+
+		tlsKeyringEntry, err := b.Core.barrier.Get(ctx, raftTLSStoragePath)
+		if err != nil {
+			return nil, err
+		}
+		if tlsKeyringEntry == nil {
+			return nil, errors.New("could not find raft TLS configuration")
+		}
+		var keyring raft.RaftTLSKeyring
+		if err := tlsKeyringEntry.DecodeJSON(&keyring); err != nil {
+			return nil, errors.New("could not decode raft TLS configuration")
+		}
+
+		if err := raftStorage.AddPeer(ctx, serverID, clusterAddr); err != nil {
+			return nil, err
+		}
+		if b.Core.raftFollowerStates != nil {
+			b.Core.raftFollowerStates.update(serverID, 0)
+		}
+
+		peers, err := raftStorage.Peers(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		return &logical.Response{
+			Data: map[string]interface{}{
+				"peers":       peers,
+				"tls_keyring": keyring,
+			},
+		}, nil
+	}
+}
+
+func (b *SystemBackend) handleStorageRaftSnapshotRead() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+		raftStorage, ok := b.Core.underlyingPhysical.(*raft.RaftBackend)
+		if !ok {
+			return logical.ErrorResponse("raft storage is not in use"), logical.ErrInvalidRequest
+		}
+		if req.ResponseWriter == nil {
+			return nil, errors.New("no writer for request")
+		}
+
+		err := raftStorage.Snapshot(req.ResponseWriter, b.Core.seal.GetAccess())
+		if err != nil {
+			return nil, err
+		}
+
+		return nil, nil
+	}
+}
+
+func (b *SystemBackend) handleStorageRaftSnapshotWrite(force bool) framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+		raftStorage, ok := b.Core.underlyingPhysical.(*raft.RaftBackend)
+		if !ok {
+			return logical.ErrorResponse("raft storage is not in use"), logical.ErrInvalidRequest
+		}
+		if req.RequestReader == nil {
+			return nil, errors.New("no writer for request")
+		}
+
+		access := b.Core.seal.GetAccess()
+		if force {
+			access = nil
+		}
+
+		// We want to buffer the http request reader into a temp file here so we
+		// don't have to hold the full snapshot in memory. We also want to do
+		// the restore in two parts so we can restore the snapshot while the
+		// stateLock is write locked.
+		snapFile, cleanup, metadata, err := raftStorage.WriteSnapshotToTemp(req.RequestReader, access)
+		switch {
+		case err == nil:
+		case strings.Contains(err.Error(), "failed to open the sealed hashes"):
+			switch b.Core.seal.BarrierType() {
+			case seal.Shamir:
+				return logical.ErrorResponse("Could not verify hash file, snapshot could be using a different set of shamir keys. Use the snapshot-force API to bypass this check."), logical.ErrInvalidRequest
```

Comment:
As a style thing, as you know normally we don't capitalize. You could do this and the one below like `could not verify hash file, possibly the snapshot is using a different set of unseal keys; use the snapshot-force API to bypass this check`